### ISSUE TITLE
Implement Proxy Support

### ISSUE THREAD

**Issue by Sann0** [ikvk/imap_tools#24]
Proxy support would be a great feature. I've found that piece of code which would add `pysocks` as a third party dependency but it might be worth it.

```python
import imaplib
import socks
import ssl

class SocksIMAP4(imaplib.IMAP4):
    PROXY_TYPES = {"socks4": socks.PROXY_TYPE_SOCKS4, "socks5": socks.PROXY_TYPE_SOCKS5, "http": socks.PROXY_TYPE_HTTP}

    def __init__(
        self,
        host: str = "",
        port: int = imaplib.IMAP4_PORT,
        proxy_addr: str = None,
        proxy_port: int = None,
        username: str = None,
        password: str = None,
        proxy_type: str = "http",
    ):

        self.proxy_addr = proxy_addr
        self.proxy_port = proxy_port
        self.username = username
        self.password = password
        self.proxy_type = SocksIMAP4.PROXY_TYPES[proxy_type.lower()]

        imaplib.IMAP4.__init__(self, host, port)

    def _create_socket(self):
        return socks.create_connection(
            (self.host, self.port),
            proxy_type=self.proxy_type,
            proxy_addr=self.proxy_addr,
            proxy_port=self.proxy_port,
            proxy_username=self.username,
            proxy_password=self.password,
        )


class SocksIMAP4SSL(SocksIMAP4):
    def __init__(
        self,
        host: str = "",
        port: int = imaplib.IMAP4_SSL_PORT,
        proxy_addr: str = None,
        proxy_port: int = None,
        username: str = None,
        password: str = None,
        proxy_type: str = "http",
    ):

        SocksIMAP4.__init__(
            self,
            host,
            port,
            proxy_addr=proxy_addr,
            proxy_port=proxy_port,
            username=username,
            password=password,
            proxy_type=proxy_type,
        )

    def _create_socket(self):
        sock = SocksIMAP4._create_socket(self)
        server_hostname = self.host if ssl.HAS_SNI else None
        return ssl._create_stdlib_context(certfile=None, keyfile=None).wrap_socket(
            sock, server_hostname=server_hostname
        )

    def open(self, host: str = "", port: int = imaplib.IMAP4_PORT):
        SocksIMAP4.open(self, host, port)```

**Comment by ikvk** [comment]
Hi, I will think about it

**Comment by thesalah1212** [comment]
we will be waiting for proxy implementation cz some IMAP server needs proxy to work

**Comment by ikvk** [comment]
Hi again.
I decided not to add support proxy (by PySocks) to imap_tools lib by now.
This task is beyond the scope of the library, and requires a lot of time to support.
Actually it is good thing and i see that people wants it.
So, I changed the lib for simple MailBox customization.
See 0.15.0 version.

And I added example: imap_tools/examples/pysocks_proxy.py
I am not tested it at all. 
I suggest you to see it and may be fix it.
